pysmt/parsing.py

```diff
@@ -0,0 +1,435 @@
+
+
+class Identifier(GrammarSymbol):
+    def __init__(self, name):
+        GrammarSymbol.__init__(self)
+        self.value = get_env().formula_manager.get_symbol(name)
+        if self.value is None:
+            raise ValueError("Undefined symbol: '%s'" % name)
+
+    def nud(self, parser):
+        return self.value
+
+
+class ExprIf(GrammarSymbol):
+    def __init__(self):
+        GrammarSymbol.__init__(self)
+        self.lbp = 5
+
+    def led(self, parser, left):
+        cond_ = left
+        then_ = parser.expression(self.lbp)
+        if type(parser.token) != ExprElse:
+            raise SyntaxError("Expected ':'")
+        parser.advance()
+        else_ = parser.expression(self.lbp)
+        return parser.mgr.Ite(cond_, then_, else_)
+
+
+class OpenPar(GrammarSymbol):
+    def __init__(self):
+        GrammarSymbol.__init__(self)
+        self.lbp = 200
+
+    def nud(self, parser):
+        r = parser.expression()
+        if type(parser.token) != ClosePar:
+            raise SyntaxError("Expected ')', got '%s'" )
+        parser.advance()
+        return r
+
+    def led(self, parser, left):
+        # a function call
+        fun = left
+        params = []
+        if type(parser.token) != ClosePar:
+            while True:
+                r = parser.expression()
+                params.append(r)
+                if type(parser.token) != ExprComma:
+                    break
+                parser.advance()
+        if type(parser.token) != ClosePar:
+            raise SyntaxError("Expected ')'")
+        parser.advance()
+        return parser.mgr.Function(fun, params)
+
+
+class OpenBrak(GrammarSymbol):
+    def __init__(self):
+        GrammarSymbol.__init__(self)
+        self.lbp = 300
+
+    def led(self, parser, left):
+        # BVExtract
+        bv = left
+        start = parser.expression()
+        if type(parser.token) != ExprElse:
+            raise SyntaxError("Expected ':'")
+        parser.advance()
+        end = parser.expression()
+        if type(parser.token) != CloseBrak:
+            raise SyntaxError("Expected ']'")
+        parser.advance()
+        return parser.mgr.BVExtract(bv,
+                                    start.constant_value(),
+                                    end.constant_value())
+
+
+class InfixOpAdapter(GrammarSymbol):
+    def __init__(self, operator, lbp):
+        GrammarSymbol.__init__(self)
+        self.operator = operator
+        self.lbp = lbp
+
+    def led(self, parser, left):
+        right = parser.expression(self.lbp)
+        return self.operator(left, right)
+
+    def __repr__(self):
+        return repr(self.operator)
+
+
+class UnaryOpAdapter(GrammarSymbol):
+    def __init__(self, operator, lbp):
+        GrammarSymbol.__init__(self)
+        self.operator = operator
+        self.lbp = lbp
+
+    def nud(self, parser):
+        right = parser.expression(self.lbp)
+        return self.operator(right)
+
+
+class Quantifier(GrammarSymbol):
```

Comment:
This is a specific Grammar Symbol, while the one before and after are "Factories".
Would it be possible to separate the "infrastructure" from the actual parsing elements?